sdk/eventhub/Azure.Messaging.EventHubs.Processor/src/EventProcessorClient.cs

```diff
@@ -935,126 +920,62 @@ private async Task RunAsync(CancellationToken cancellationToken)
-                        .ToList();
-                    Logger.CompleteOwnershipRecordsCount(completeOwnershipList.Count());
+                    partitionIds = await consumer.GetPartitionIdsAsync(cancellationToken).ConfigureAwait(false);
                 }
                 catch (Exception ex)
                 {
                     cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
 
-                    // If ownership list retrieval fails, give up on the current cycle.  There's nothing more we can do
-                    // without an updated ownership list.
-
-                    Logger.ListOwnershipAsyncError(FullyQualifiedNamespace, EventHubName, ConsumerGroup, ex.Message);
-                    var errorEventArgs = new ProcessErrorEventArgs(null, Resources.OperationListOwnership, ex, cancellationToken);
+                    var errorEventArgs = new ProcessErrorEventArgs(null, Resources.OperationGetPartitionIds, ex, cancellationToken);
                     _ = OnProcessErrorAsync(errorEventArgs);
                 }
 
-                // There's no point in continuing the current cycle if we failed to fetch the completeOwnershipList.
+                // There's no point in continuing the current cycle if we failed to fetch the partitionIds.
 
-                if (completeOwnershipList != default)
+                if (partitionIds != default)
                 {
-                    // Get a complete list of the partition ids present in the Event Hub.  This should be immutable for the time being, but
-                    // it may change in the future.
-
-                    var partitionIds = default(string[]);
+                    PartitionOwnership claimedOwnership = default;
 
                     try
                     {
-                        partitionIds = await consumer.GetPartitionIdsAsync(cancellationToken).ConfigureAwait(false);
+                        claimedOwnership = await LoadBalancer.RunLoadBalancingAsync(partitionIds, cancellationToken);
+                    }
+                    catch (EventHubsException ex)
+                    {
+                        var errorEventArgs = new ProcessErrorEventArgs(null, ex.Message, ex, cancellationToken);
+                        _ = OnProcessErrorAsync(errorEventArgs);
                     }
                     catch (Exception ex)
                     {
-                        cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
-
-                        Logger.GetPartitionIdsAsyncError(FullyQualifiedNamespace, EventHubName, ConsumerGroup, ex.Message);
-                        var errorEventArgs = new ProcessErrorEventArgs(null, Resources.OperationGetPartitionIds, ex, cancellationToken);
+                        var errorEventArgs = new ProcessErrorEventArgs(null, string.Empty /*which resource to use here?*/, ex, cancellationToken);
                         _ = OnProcessErrorAsync(errorEventArgs);
                     }
```

Comment:
Not sure that I'm following.  If I'm reading correctly, `RunAsync` is only called during start, but isn't awaited immediately.   Since the call for getting partitions is async, by the time we hit this block, there's no chance that we're running synchonously.... so throwing here wouldn't bubble to the start call and the processor would be in an unknown invalid state.    The desired behavior would be to either refuse to start entirely or trigger the handler and make every reasonable attempt to soldier-on.

I feel like I must be overlooking something.   What am I missing?

Reply:
We'd probably want to log, though... since we're not in a scenario where it's a known Event Hubs-related operation that would have already done so.

Reply:
Apologies if you already know what I'm about to explain here, @jsquire. It's been a month since you gave this PR a full pass, so I'm assuming it's better to set the context properly.

If I recall correctly, in the processor we have three exception scenarios:
1. Exceptions that can be surfaced immediately to the user won't be logged and we simply throw them. This could happen in `StartProcessing` or in `UpdateCheckpoint`, for example. We don't catch them because these methods do not run as separate tasks.
2. Silent service exceptions we know could happen are caught and we keep the processor alive (by "silent", I mean "thrown in a background task"). We pass the caught exception to `ProcessError` with an `Operation` string to identify which task was being performed when it happened. Accessing the `StorageManager` or the Event Hubs service usually falls under this scenario. 
3. Every other silent exception is not being caught by us and they make the processor stop running. They are also not being logged. The user will only be able to know the processor status via the `IsRunning` property or by calling `StopProcessing` to catch the exception. We cannot avoid these unless we encapsulate all processing in a try/catch.

When the `PartitionLoadBalancer` class was created, some of the service exceptions in scenario 2 could now happen in the new class. This is not good because the load balancer doesn't have the error handler, whereas the event processor doesn't have the context to figure out which `Operation` string to use.

The approach @christothes took was enclosing the `Operation` string in an `EventHubsException` message. Here's an example:

https://github.com/Azure/azure-sdk-for-net/blob/fe5f340d621ce6f6d580b5a5d0c41a48c8db7ea8/sdk/eventhub/Azure.Messaging.EventHubs.Shared/src/Processor/PartitionLoadBalancer.cs#L155-L163

This exception will be caught by the event processor and the error handler will be called accordingly.

What I'm trying to say is: if the processor catches an exception that it's not an `EventHubsException`, it means it's an exception that falls under the third scenario. For this reason there is no `Operation` string to use, because they are only for exceptions in scenario 2.

I think it's reasonable to treat it as any other scenario 3 exception and don't catch it. Is there any reason to treat them differently? If you think it's important to log it, instead of catching it here we should enclose the whole `Run` method in a try/catch block so we can catch and log all exceptions that fall under scenario 3.

Reply:
Thanks for the context;  it has been a while and I've been spread around.  In general, #3 should be considered either a catastrophic scenario (out of memory, thread abort, alien invasion) or a failing of anticipation on our part.    We're striving to have everything fall into bucket #2 that we possibly can.    So, if I understand correctly in this context, thanks to the catch that @christothes added, we've got a blanket rewrap of load balancer exceptions as `EventHubsException` instances marked as transient.   Therefore, if something is coming in as any other type, then its not within the load balancer and we don't have an understanding of what caused the issue.

I'd still pose the question, however, of whether the scenario of an unexpected exception puts us at risk of data corruption, being in an invalid state from which we cannot recover, or would otherwise be a risk by treating as recoverable.   Since we're not awaiting the exception handler, there is no user code that would potentially trigger an exception - so this would be something internal to the processor itself.   Given the scope of the `try/catch` it would be something within the load balancer that didn't hit a `catch` block.    I don't see this as a major concern, assuming that we're confident that the load balancer is wrapping unconditionally.    However, I would still prefer that we log either in this scope of an upper level scope so that we can at least capture information about unhandled and unexpected things.

Reply:
logged the exception